### PR TITLE
[SPIRV] Implement support for SPV_KHR_expect_assume

### PR DESCRIPTION
**Pull request by pmatos**
Adds new extension SPV_KHR_expect_assume, new capability
ExpectAssumeKHR as well as the new instructions:
  * OpExpectKHR
  * OpAssumeTrueKHR

These are lowered from respectively llvm.expect.<ty> and llvm.assume
intrinsics.

Previously https://reviews.llvm.org/D157696

### CODE REVIEW

**Comment by pmatos** [comment]
Still need to address https://reviews.llvm.org/D157696#4614823

File: llvm/lib/Target/SPIRV/SPIRVBuiltins.td [hunk 1]
```diff
@@ -54,6 +54,7 @@ def Enqueue : BuiltinGroup;
 def AsyncCopy : BuiltinGroup;
 def VectorLoadStore : BuiltinGroup;
 def LoadStore : BuiltinGroup;
+def ExpectAssume : BuiltinGroup;
```
Comment (iliya-diyachkov) [inline llvm/lib/Target/SPIRV/SPIRVBuiltins.td:57]: I'm not sure where this new ExpectAssume is actually used. If you are going to use ExpectAssume builtin in later version of the patch (or in the next patch), please add it later.

File: llvm/lib/Target/SPIRV/SPIRVInstrInfo.td [hunk 2]
```diff
@@ -757,3 +757,7 @@ def OpGroupNonUniformBitwiseXor: OpGroupNUGroup<"BitwiseXor", 361>;
 def OpGroupNonUniformLogicalAnd: OpGroupNUGroup<"LogicalAnd", 362>;
 def OpGroupNonUniformLogicalOr: OpGroupNUGroup<"LogicalOr", 363>;
 def OpGroupNonUniformLogicalXor: OpGroupNUGroup<"LogicalXor", 364>;
+
+// SPV_KHR_expect_assume : Expect assume instructions
+def OpAssumeTrueKHR: Op<5630, (outs), (ins ID:$cond), "OpAssumeTrueKHR $cond">;
+def OpExpectKHR: Op<5631, (outs ID:$res), (ins TYPE:$ty, ID:$val, ID:$expected), "$res = OpExpectKHR $ty $val $expected">;
```
Comment (iliya-diyachkov) [inline llvm/lib/Target/SPIRV/SPIRVInstrInfo.td:763]: In SPIR-V Specification (v 1.6), these instructions are in the section "3.42.1. Miscellaneous Instructions". Shouldn't we put the defs in the corresponding section too (i.e. after "def OpSizeOf")?

File: llvm/lib/Target/SPIRV/SPIRVModuleAnalysis.cpp [hunk 3]
```diff
@@ -881,6 +883,11 @@ void addInstrRequirements(const MachineInstr &MI,
   case SPIRV::OpGroupNonUniformBallotFindMSB:
     Reqs.addCapability(SPIRV::Capability::GroupNonUniformBallot);
     break;
+  case SPIRV::OpAssumeTrueKHR:
+  case SPIRV::OpExpectKHR:
+    Reqs.addExtension(SPIRV::Extension::SPV_KHR_expect_assume);
+    Reqs.addCapability(SPIRV::Capability::ExpectAssumeKHR);
```
Comment (iliya-diyachkov) [inline llvm/lib/Target/SPIRV/SPIRVModuleAnalysis.cpp:889]: Don't we need to check ST.canUseExtension()?

File: llvm/lib/Target/SPIRV/SPIRVInstrInfo.td (hunk 2 above)
Comment (pmatos) [reply llvm/lib/Target/SPIRV/SPIRVInstrInfo.td:763]: You're right - moving them to the right place. Thanks.

File: llvm/lib/Target/SPIRV/SPIRVSubtarget.cpp [hunk 4]
```diff
@@ -11,6 +11,7 @@
 //===----------------------------------------------------------------------===//
 
 #include "SPIRVSubtarget.h"
+#include "MCTargetDesc/SPIRVBaseInfo.h"
```
Comment (iliya-diyachkov) [inline llvm/lib/Target/SPIRV/SPIRVSubtarget.cpp:14]: Do we really need this?

**Review by iliya-diyachkov — COMMENTED** [review]
In general it looks good, but I found a few small issues.

File: llvm/lib/Target/SPIRV/SPIRVSubtarget.cpp (hunk 4 above)
Comment (pmatos) [reply llvm/lib/Target/SPIRV/SPIRVSubtarget.cpp:14]: Added this initially for another reason and it's not needed any longer. Thanks.

File: llvm/test/CodeGen/SPIRV/assume.ll [hunk 5]
```diff
@@ -0,0 +1,15 @@
+; RUN: llc -mtriple=spirv32-unknown-unknown < %s | FileCheck %s
+; RUN: llc -mtriple=spirv64-unknown-unknown < %s | FileCheck %s
+
+; CHECK:      OpCapability ExpectAssumeKHR
+; CHECK-NEXT: OpExtension "SPV_KHR_expect_assume"
+
+declare void @llvm.assume(i1)
+
+;  CHECK-DAG: %9 = OpIEqual %5 %6 %7
+; CHECK-NEXT: OpAssumeTrueKHR %9
```
Comment (iliya-diyachkov) [inline llvm/test/CodeGen/SPIRV/assume.ll:10]: Also left alignment...

File: llvm/test/CodeGen/SPIRV/expect.ll [hunk 6]
```diff
@@ -0,0 +1,20 @@
+; RUN: llc -mtriple=spirv32-unknown-unknown < %s | FileCheck %s
+; RUN: llc -mtriple=spirv64-unknown-unknown < %s | FileCheck %s
+
+;      CHECK: OpCapability ExpectAssumeKHR
+; CHECK-NEXT: OpExtension "SPV_KHR_expect_assume"
```
Comment (iliya-diyachkov) [inline llvm/test/CodeGen/SPIRV/expect.ll:5]: Also left alignment...

**Comment by pmatos** [comment]
I have rebased on main to resolve conflicts with the just landed SPV_KHR_bit_instructions support. Force-pushed branch.

**Review by iliya-diyachkov — APPROVED** [review]
It looks good to me. Thanks for your patch, Paulo.

**Comment by pmatos** [comment]
> It looks good to me. Thanks for your patch, Paulo.

Thanks for your patience reviewing this.

File: llvm/include/llvm/IR/IntrinsicsSPIRV.td [hunk 7]
```diff
@@ -33,4 +33,8 @@ let TargetPrefix = "spv" in {
   def int_spv_unreachable : Intrinsic<[], []>;
   def int_spv_alloca : Intrinsic<[llvm_any_ty], []>;
   def int_spv_undef : Intrinsic<[llvm_i32_ty], []>;
-}
+
+  // Expect, Assume Intrinsics
+  def int_spv_assume : Intrinsic<[], [llvm_i1_ty]>;
+  def int_spv_expect : Intrinsic<[llvm_anyint_ty], [LLVMMatchType<0>, LLVMMatchType<0>]>;
+}
```
Comment (michalpaszkowski) [inline llvm/include/llvm/IR/IntrinsicsSPIRV.td:40]: Please add a new line at the end of file.

File: llvm/test/CodeGen/SPIRV/assume.ll [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+declare void @llvm.assume(i1)
+
+; CHECK-DAG:  %9 = OpIEqual %5 %6 %7
+; CHECK-NEXT: OpAssumeTrueKHR %9
+define void @assumeeq(i32 %x, i32 %y) {
+    %cmp = icmp eq i32 %x, %y
+    call void @llvm.assume(i1 %cmp)
+    ret void
+}
```
Comment (michalpaszkowski) [inline llvm/test/CodeGen/SPIRV/assume.ll:15]: Please add a new line at the end of file.

**Review by michalpaszkowski — APPROVED** [review]
Thank you @pmatos for the patch and thank you @iliya-diyachkov for the comments! The patch LGTM!

One last thing before you merge, please add the missing lines at the end of `IntrinsicsSPIRV.td` and `assume.ll` files for consistency.

**Comment by pmatos** [comment]
> Thank you @pmatos for the patch and thank you @iliya-diyachkov for the comments! The patch LGTM!
> 
> One last thing before you merge, please add the missing lines at the end of `IntrinsicsSPIRV.td` and `assume.ll` files for consistency.

thanks for the remaining comments. just in time before merging. :)

**Pull request merged**